url2df0f7c811_przemekgalazka_legacycode_tgz-pJ8-pl_cc_core_cmd_CommandJ8/src/pl/cc/utils/MD5Crypt.java

```diff
@@ -0,0 +1,342 @@
+* this stuff is worth it, you can buy me a beer in return.   Poul-Henning Kamp
+* ----------------------------------------------------------------------------
+* </pre></p>
+*/
+
+public final class MD5Crypt {
+
+/**
+*
+* Command line test rig.
+ * @throws NoSuchAlgorithmException 
+*
+*/
+
+static public void main(String argv[]) throws NoSuchAlgorithmException
+{
+ 
+	 System.err.println(MD5Crypt.crypt("aga123", "$1$sk]WAINt$ig9k.5XzVesVNrtH5uUFj."));
+	 System.err.println(MD5Crypt.crypt("bąk", "$1$:Ch]iqPw$5ZvkDlNfyZFliDN0XCaQC0"));
+ 
+ System.exit(0);
+}
+
+static private final String SALTCHARS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890";
+
+static private final String itoa64 = "./0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+
+static private final String to64(long v, int size)
+{
+ StringBuffer result = new StringBuffer();
+
+ while (--size >= 0)
+   {
+	result.append(itoa64.charAt((int) (v & 0x3f)));
+	v >>>= 6;
+   }
+
+ return result.toString();
+}
+
+static private final void clearbits(byte bits[])
+{
+ for (int i = 0; i < bits.length; i++)
+   {
+	bits[i] = 0;
+   }
+}
+
+
+
+/**
+* convert an encoded unsigned byte value into a int
+* with the unsigned value.
+*/
+
+static private final int bytes2u(byte inp)
+{
+ return (int) inp & 0xff;
+}
+
+/**
+* <p>This method actually generates a OpenBSD/FreeBSD/Linux PAM compatible
+* md5-encoded password hash from a plaintext password and a
+* salt.</p>
+*
+* <p>The resulting string will be in the form '$1$&lt;salt&gt;$&lt;hashed mess&gt;</p>
+*
+* @param password Plaintext password
+*
+* @return An OpenBSD/FreeBSD/Linux-compatible md5-hashed password field.
+ * @throws NoSuchAlgorithmException 
+*/
+
+static public final String crypt(String password) throws NoSuchAlgorithmException
+{
+ StringBuffer salt = new StringBuffer();
+ java.util.Random randgen = new java.util.Random();
+
+ /* -- */
+
+  while (salt.length() < 8)
+    {
+	 int index = (int) (randgen.nextFloat() * SALTCHARS.length());
+      salt.append(SALTCHARS.substring(index, index+1));
+    }
+
+  return MD5Crypt.crypt(password, salt.toString());
+}
+
+/**
+* <p>This method actually generates a OpenBSD/FreeBSD/Linux PAM compatible
+* md5-encoded password hash from a plaintext password and a
+* salt.</p>
+*
+* <p>The resulting string will be in the form '$1$&lt;salt&gt;$&lt;hashed mess&gt;</p>
+*
+* @param password Plaintext password
+* @param salt A short string to use to randomize md5.  May start with $1$, which
+*             will be ignored.  It is explicitly permitted to pass a pre-existing
+*             MD5Crypt'ed password entry as the salt.  crypt() will strip the salt
+*             chars out properly.
+*
+* @return An OpenBSD/FreeBSD/Linux-compatible md5-hashed password field.
+ * @throws NoSuchAlgorithmException 
+*/
+
+static public final String crypt(String password, String salt) throws NoSuchAlgorithmException
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 232-262 into a separate method.